src/python/pants/core/util_rules/adhoc_process_support.py

```diff
@@ -62,45 +72,144 @@ class AdhocProcessResult:
 @dataclass(frozen=True)
 class ResolveExecutionDependenciesRequest:
     address: Address
-    execution_dependencies: tuple[str, ...] | None
+    execution_dependencies: tuple[str | runnable, ...] | None
     dependencies: tuple[str, ...] | None  # can go away after 2.17.0.dev0 per deprecation
 
 
 @dataclass(frozen=True)
 class ResolvedExecutionDependencies:
     digest: Digest
+    runnable_dependencies: RunnableDependencies | None
+
+
+@dataclass(frozen=True)
+class runnable:
+    """Allows for an execution dependency to be specified as `runnable`.
+
+    When resolved, the dependency will be available on the `PATH` with the name `name`.
+
+    `runnable` dependencies are not resolved transitively, and must be defined on the target that
+    wants to run those dependencies.
+    """
+
+    name: str
+    address: str
```

Comment:
This is an interesting approach, I'd like to hear @thejcannon and @kaos's opinion on it.

I think that in order for this field to be accounted for in `./pants {dependencies,dependents}`, `--changed`, etc, this field type will need to actually be reified into dependencies the way that `SpecialCasedDependencies` is by `DependenciesRequest`: https://github.com/pantsbuild/pants/blob/38c6c0ed31736f6ef271374d51cdbc4c7df1afa4/src/python/pants/engine/internals/graph.py#L1253-L1265

One way to use a non-standard structure like this but still actually have dependencies tracked would be to replace the strategy used by `SpecialCasedDependencies` (which currently involves subclassing), to one which is based on a `@union`: then you could install a `@rule` (or declare it as a method of the `@union` base class thanks to @kaos's work on `@rule_helper`s) that would turn your dependencies field type in to a flat list of dependencies as expected by `DependenciesRequest`. I actually think that this could be a fairly easy change to make!

Having said all that: I'm not sure about the wrapping `object` mixed into another field vs e.g., using a `dict[Name,Address]`. The above strategy would allow either type to be used to declare dependencies. IMO, having different dependencies fields with different documentation is a good thing: to the extent

Reply:
Thanks for explaining the real value of `SpecialCasedDependencies`. I couldn't see how it differed from a `StringSequenceField`, but your comment does a good job of explaining it. Thanks!

Currently `SpecialCasedDependencies` subclasses `StringSequenceField`, so that's a thing we'd need to change, so yes, doing something based on Unions might be a good approach. I'm not certain that adding _even more_ dependency fields are good here as far as prior art goes.

I _think_ using a custom object is a better approach than using a dict for documentation/semantics but I could be wrong.

Reply:
@stuhood So if we made `execution_dependencies` trackable in the same way as `SpecialCasedDependencies`, would we be happy enough with following the proposed API for the time being? I'm leaning towards keeping the custom object approach for the purposes of type checking.

Reply:
Honestly, I think that @kaos's suggestion in the other thread to use the target name (in a separate dependencies field) is workable for now, and punts on the (big) question of how to add edge/dependency labels.

I agree that we should tackle that problem, but the status quo right now is: "use another `(SpecialCased)Dependencies`" field, and trying to introduce a new paradigm a few hours before we cut a stable branch doesn't seem like the best time.

Reply:
Fair enough.